tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/ScalaJSOptimizer.scala

```diff
@@ -1,285 +0,0 @@
-
-    CacheUtils.cached(classpath.version, cfg.output, cfg.cache) {
-      logger.info(s"Fast optimizing ${cfg.output.path}")
-      optimizeIR(classpath.scalaJSIR, cfg, logger)
-    }
-
-    new LinkedClasspath(classpath.jsLibs, cfg.output,
-        classpath.requiresDOM, classpath.version)
-  }
-
-  def optimizeIR(irFiles: Traversable[VirtualScalaJSIRFile],
-      cfg: Config, logger: Logger): Unit = {
-
-    val builder = {
-      import cfg._
-      if (wantSourceMap)
-        new JSFileBuilderWithSourceMap(output.name,
-            output.contentWriter,
-            output.sourceMapWriter,
-            relativizeSourceMapBase)
-      else
-        new JSFileBuilder(output.name, output.contentWriter)
-    }
-
-    emitter.emitCustomHeader(cfg.customOutputWrapper._1, builder)
-    emitter.emitPrelude(builder, logger)
-    optimizeIR(irFiles, cfg, builder, logger)
-    emitter.emitPostlude(builder, logger)
-    emitter.emitCustomFooter(cfg.customOutputWrapper._2, builder)
-
-    builder.complete()
-    builder.closeWriters()
-  }
-
-  def optimizeIR(irFiles: Traversable[VirtualScalaJSIRFile],
-      cfg: OptimizerConfig, builder: JSTreeBuilder, logger: Logger): Unit = {
-
-    // Update state that is a per-run configuration for the overall optimizer,
-    // but not the phases
-    if (cfg.wantSourceMap != withSourceMap) {
-      withSourceMap = cfg.wantSourceMap
-      resetState()
-    }
-
-    val linkResult = try {
-      logTime(logger, "Linker") {
-        linker.link(irFiles, logger,
-            reachOptimizerSymbols = !cfg.disableOptimizer,
-            cfg.bypassLinkingErrors, cfg.checkIR)
-      }
-    } catch {
-      case th: Throwable =>
-        resetState()
-        throw th
-    }
-
-    val useOptimizer = linkResult.isComplete && !cfg.disableOptimizer
-
-    if (cfg.batchMode)
-      resetStateFromOptimizer()
-
-    try {
-      val finalResult = if (useOptimizer) {
-        val rawOptimized = logTime(logger, "Inc. optimizer") {
-          optimizer.update(linkResult, logger)
-        }
-
-        logTime(logger, "Refiner") {
-          refiner.refine(rawOptimized, logger)
-        }
-      } else {
-        if (!cfg.disableOptimizer)
-          logger.warn("Not running the optimizer because there where linking errors.")
-        linkResult
-      }
-
-      logTime(logger, "Emitter (write output)") {
-        emitter.emit(finalResult, builder, logger)
-      }
-    } catch {
-      case th: Throwable =>
-        resetStateFromOptimizer()
-        throw th
-    }
-  }
-
-  /** Resets all persistent state of this optimizer */
-  def clean(): Unit = resetState()
-
-  private def resetState(): Unit = {
-    linker = new Linker(semantics, outputMode, withSourceMap)
-    resetStateFromOptimizer()
-  }
-
-  private def resetStateFromOptimizer(): Unit = {
-    optimizer = optimizerFactory(semantics, outputMode, withSourceMap)
-    refiner = new Refiner(semantics, outputMode)
-    emitter = new Emitter(semantics, outputMode)
```

Comment:
Yes. The `Linker` protects itself from being reused if it throws an exception, the lower-level classes may not be reused by spec. The `ClearableLinker` recreates its internal linker if it throws or if it is requested.